fdbclient/include/fdbclient/BlobGranuleRequest.actor.h

```diff
@@ -0,0 +1,186 @@
+ * limitations under the License.
+ */
+
+#pragma once
+
+// When actually compiled (NO_INTELLISENSE), include the generated version of this file.  In intellisense use the source
+// version.
+#if defined(NO_INTELLISENSE) && !defined(FDBCLIENT_BLOB_GRANULE_REQUEST_ACTOR_G_H)
+#define FDBCLIENT_BLOB_GRANULE_REQUEST_ACTOR_G_H
+#include "fdbclient/BlobGranuleRequest.actor.g.h"
+#elif !defined(FDBCLIENT_BLOB_GRANULE_REQUEST_ACTOR_H)
+#define FDBCLIENT_BLOB_GRANULE_REQUEST_ACTOR_H
+
+#include "flow/flow.h"
+#include "flow/Knobs.h"
+
+// #include "fdbclient/NativeAPI.actor.h"
+#include "flow/Arena.h"
+#include "fdbclient/DatabaseContext.h"
+#include "fdbclient/BlobWorkerInterface.h"
+
+#include "flow/actorcompiler.h" // This must be the last #include.
+
+#define BGR_DEBUG false
+
+ACTOR template <class Request, bool P>
+Future<Standalone<VectorRef<REPLY_TYPE(Request)>>> txnDoBlobGranuleRequests(
+    Transaction* tr,
+    Key* beginKey,
+    Key endKey,
+    Request request,
+    RequestStream<Request, P> BlobWorkerInterface::*channel) {
+	// TODO KNOB
+	state RangeResult blobGranuleMapping = wait(krmGetRanges(
+	    tr, blobGranuleMappingKeys.begin, KeyRangeRef(*beginKey, endKey), 64, GetRangeLimits::BYTE_LIMIT_UNLIMITED));
+
+	state int i = 0;
+	state std::vector<Future<ErrorOr<REPLY_TYPE(Request)>>> requests;
+	state Standalone<VectorRef<REPLY_TYPE(Request)>> results;
+
+	for (; i < blobGranuleMapping.size() - 1; i++) {
+		if (!blobGranuleMapping[i].value.size()) {
+			if (BGR_DEBUG) {
+				fmt::print("ERROR: No valid granule data for range [{0} - {1}) \n",
+				           blobGranuleMapping[i].key.printable(),
+				           blobGranuleMapping[i + 1].key.printable());
+			}
+			// no granule for range
+			throw blob_granule_transaction_too_old();
+		}
+
+		state UID workerId = decodeBlobGranuleMappingValue(blobGranuleMapping[i].value);
+		if (workerId == UID()) {
+			if (BGR_DEBUG) {
+				fmt::print("ERROR: Invalid Blob Worker ID for range [{0} - {1}) \n",
+				           blobGranuleMapping[i].key.printable(),
+				           blobGranuleMapping[i + 1].key.printable());
+			}
+			// no worker for granule
+			throw blob_granule_transaction_too_old();
+		}
+
+		if (!tr->trState->cx->blobWorker_interf.count(workerId)) {
+			Optional<Value> workerInterface = wait(tr->get(blobWorkerListKeyFor(workerId)));
+			// from the time the mapping was read from the db, the associated blob worker
+			// could have died and so its interface wouldn't be present as part of the blobWorkerList
+			// we persist in the db.
+			if (workerInterface.present()) {
+				tr->trState->cx->blobWorker_interf[workerId] = decodeBlobWorkerListValue(workerInterface.get());
+			} else {
+				if (BGR_DEBUG) {
+					fmt::print("ERROR: Worker  for range [{1} - {2}) does not exist!\n",
+					           workerId.toString().substr(0, 5),
+					           blobGranuleMapping[i].key.printable(),
+					           blobGranuleMapping[i + 1].key.printable());
+				}
+				break;
```

Comment:
if there is no worker for certain range, it means we couldn't flush this range and the data is still in-memory.. we should not proceed to truncate logs, right? if so, we may need to throw the error and fail the flush?

Reply:
So the interaction here is a bit complicated.
The main case that causes the granule mapping to say that a key range belongs to a worker, but that worker doesn't exist, is if that worker died. This means that the blob manager should already be in the process of assigning that granule to a new, alive worker.
High-level, what we want to do here is just wait and retry. How we actually need to do this is to get a higher read version for this loop, so that we can read the newer granule mapping and see where the new worker is.

So to accomplish that, we break out of this loop, and return to the caller.

It might be better to explicitly throw transaction_too_old or something that then forces the caller to get a newer read version instead of waiting several seconds for the old one to get transaction_too_old though